core/server/worker/src/main/java/alluxio/worker/netty/DataServerBlockReadHandler.java

```diff
@@ -85,12 +94,36 @@ public void close() throws IOException {
    */
   DataServerBlockReadHandler(ExecutorService executorService, BlockWorker blockWorker,
       FileTransferType fileTransferType) {
-    super(executorService);
+    super(executorService, blockWorker);
     mWorker = blockWorker;
     mTransferType = fileTransferType;
   }
 
   @Override
+  public void channelUnregistered(ChannelHandlerContext ctx) {
+    if (mRequest != null) {
+      try {
+        mRequest.close();
+      } catch (IOException e) {
+        LOG.warn("Failed to close block read request {} with error {}.", mRequest, e.getMessage());
+      }
+    }
+    ctx.fireChannelUnregistered();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+    if (mRequest != null) {
+      try {
+        mRequest.close();
```

Comment:
This can introduce races. The PacketReader thread could be using mRequest to read data. After creation, only the PacketReader thread can mutate the state of this. I think the current code closes mRequest upon failure properly.